HARK/distribution.py

```diff
@@ -0,0 +1,538 @@
+        probability.  Should be >= 1 for sensible spacing.
+
+    Returns
+    -------
+    d : DiscreteDistribution
+        Probability associated with each point in array of discrete
+        points for discrete probability mass function.
+
+    Written by Luca Gerotto
+    Based on Matab function "setup_workspace.m," from Chris Carroll's
+      [Solution Methods for Microeconomic Dynamic Optimization Problems]
+      (http://www.econ2.jhu.edu/people/ccarroll/solvingmicrodsops/) toolkit.
+    Latest update: 11 February 2017 by Matthew N. White
+    '''
+    # Find the CDF boundaries of each segment
+    if sigma > 0.0:
+        if tail_N > 0:
+            lo_cut     = tail_bound[0]
+            hi_cut     = tail_bound[1]
+        else:
+            lo_cut     = 0.0
+            hi_cut     = 1.0
+        inner_size     = hi_cut - lo_cut
+        inner_CDF_vals = [lo_cut + x*N**(-1.0)*inner_size for x in range(1, N)]
+        if inner_size < 1.0:
+            scale      = 1.0/tail_order
+            mag        = (1.0-scale**tail_N)/(1.0-scale)
+        lower_CDF_vals = [0.0]
+        if lo_cut > 0.0:
+            for x in range(tail_N-1,-1,-1):
+                lower_CDF_vals.append(lower_CDF_vals[-1] + lo_cut*scale**x/mag)
+        upper_CDF_vals  = [hi_cut]
+        if hi_cut < 1.0:
+            for x in range(tail_N):
+                upper_CDF_vals.append(upper_CDF_vals[-1] + (1.0-hi_cut)*scale**x/mag)
+        CDF_vals       = lower_CDF_vals + inner_CDF_vals + upper_CDF_vals
+        temp_cutoffs   = list(stats.lognorm.ppf(CDF_vals[1:-1], s=sigma, loc=0, scale=np.exp(mu)))
+        cutoffs        = [0] + temp_cutoffs + [np.inf]
+        CDF_vals       = np.array(CDF_vals)
+
+        # Construct the discrete approximation by finding the average value within each segment.
+        # This codeblock ignores warnings because it throws a "divide by zero encountered in log"
+        # warning due to computing erf(infty) at the tail boundary.  This is irrelevant and
+        # apparently freaks new users out.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            K              = CDF_vals.size-1 # number of points in approximation
+            pmf            = CDF_vals[1:(K+1)] - CDF_vals[0:K]
+            X              = np.zeros(K)
+            for i in range(K):
+                zBot  = cutoffs[i]
+                zTop = cutoffs[i+1]
+                tempBot = (mu+sigma**2-np.log(zBot))/(np.sqrt(2)*sigma)
+                tempTop = (mu+sigma**2-np.log(zTop))/(np.sqrt(2)*sigma)
+                if tempBot <= 4:
+                    X[i] = -0.5*np.exp(mu+(sigma**2)*0.5)*(erf(tempTop) - erf(tempBot))/pmf[i]
+                else:
+                    X[i] = -0.5*np.exp(mu+(sigma**2)*0.5)*(erfc(tempBot) - erfc(tempTop))/pmf[i]
+
+    else:
+        pmf = np.ones(N)/N
+        X   = np.exp(mu)*np.ones(N)
+    return DiscreteDistribution(pmf, X)
+
+@memoize
```

Comment:
Any ideas why is this here?
(I know this isn't this PR change)

Reply:
Are you referring to the @memoize decorator?
I assumed it was for performance reasons.